criu/pages-compress.c

```diff
@@ -0,0 +1,285 @@
+	char pathdst[PATH_MAX];
+	const long pid = getpid();
+	const int dfd = get_service_fd(IMG_FD_OFF);
+	const int id = 1; // the only image we'd like to compress
+	int ret;
+
+	LZ4IO_prefs_t *lz4_prefs = LZ4IO_defaultPreferences();
+	LZ4IO_setBlockSize(lz4_prefs, 128 * 1024);
+
+	pr_debug("Compressing pages\n");
+
+	snprintf(format, sizeof(format), "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES].fmt);
+	snprintf(pathsrc, sizeof(pathsrc), format, pid, dfd, id);
+
+	snprintf(format, sizeof(format), "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES_COMP].fmt);
+	snprintf(pathdst, sizeof(pathdst), format, pid, dfd, id);
+
+	ret = LZ4IO_compressFilename(pathsrc, pathdst, 8, lz4_prefs);
+	LZ4IO_freePreferences(lz4_prefs);
+
+	if (ret) {
+		pr_err("Can't compress %s to %s\n", pathsrc, pathdst);
+		return -1;
+	}
+	if (unlink(pathsrc)) {
+		pr_perror("Can't delete uncompressed source pages image: %s\n", pathsrc);
+		return -1;
+	}
+	return 0;
+}
+
+static int decompress_image3(const char *mapped_addr, size_t mapped_size, LZ4F_decompressionContext_t dctx) {
+	const char *compbuf = mapped_addr;
+	const int outbufsize = 4 * 1024 * 1024; // LZ4F_max4MB is the max block size; see LZ4F_blockSizeID_t for details
+	char outbuf[outbufsize];
+	LZ4F_errorCode_t lz4err;
+	int decomp_fd;
+	size_t totalread = 0;
+	size_t totalwrite = 0;
+
+	snprintf(decompressed_filename, sizeof(decompressed_filename), decompressed_filename_format, (int)getpid());
+	pr_debug("Creating tmp file at '%s'\n", decompressed_filename);
+	decomp_fd = open(decompressed_filename, O_CREAT | O_WRONLY | O_TRUNC, S_IRUSR | S_IWUSR);
+
+	if (0 > decomp_fd) {
+		pr_err("Can't create file, errno=%d\n", errno);
+		return -1;
+	}
+
+	while (1) {
+		size_t insize = mapped_size - (compbuf - mapped_addr);
+		if (!insize) {
+			/* reached end of file or stream, no data left to decompress */
+			decompressionResult = 0;
+			break;
+		}
+		{
+			size_t outsize = sizeof outbuf;
+			lz4err = LZ4F_decompress(dctx, outbuf, &outsize, compbuf, &insize, NULL);
+			if (LZ4F_isError(lz4err)) {
+				pr_err("LZ4 Decompress error: %s\n", LZ4F_getErrorName(lz4err));
+				break;
+			}
+
+			totalread += insize;
+			compbuf += insize;
+			if (mapped_size < (compbuf - mapped_addr)) {
+				pr_err("decompression went wrong, compression buffer is out of bounds\n");
+				break;
+			}
+
+			{
+				// Write decompressed bytes
+				ssize_t bytes_written = 0;
+				ssize_t out_offset = 0;
+				do {
+					bytes_written = write(decomp_fd, outbuf + out_offset, outsize);
+					if (0 > bytes_written) {
+						if (EINTR == errno) {
+							continue;
+						}
+						pr_perror ("write error to decompressed file");
+						break;
+					}
+					out_offset += bytes_written;
+					if (outsize < (size_t)bytes_written) {
+						pr_err("written more bytes in total than needed, expected=%lu, written=%lu\n",
+						       (unsigned long)outsize, (unsigned long)bytes_written);
+						break;
+					}
+					outsize -= bytes_written;
```

Comment:
A nit: I'd prefer these two variables updates to be closer
```
out_offset += bytes_written;
outsize -= bytes_written;
```
I cannot image `write` to write more bytes than was requested, the `outsize < (size_t)bytes_written` test can be `assert`ed.